src/main/groovy/com/lesfurets/jenkins/unit/DeclarativePipelineHarness.groovy

```diff
@@ -0,0 +1,164 @@
+
+        registerCallStack('agent')
+        registerCallStack('docker')
+        registerCallStack('dockerfile')
+        registerCallStack('node')
+
+        registerCallStack('stages')
+        registerCallStack('parallel')
+        registerCallStack('script')
+        registerCallStack('steps')
+
+        registerCallStack('when')
+        registerCallStack('allOf')
+        registerCallStack('anyOf')
+        registerCallStack('expression')
+
+        //Register post methods
+        registerCallStack('post')
+        registerCallStack('always')
+        registerCallStack('cleanup')
+        registerCallStack('success')
+        registerCallStack('failure')
+        registerCallStack('regression')
+        registerCallStack('changed')
+        registerCallStack('fixed')
+        registerCallStack('aborted')
+        registerCallStack('unstable')
+        registerCallStack('unsuccessful')
+
+        registerIdentifiedCallStack('stage', String.class)
+        registerIdentifiedCallStack('node', String.class)
+        registerIdentifiedCallStack('withCredentials', Object.class)
+        registerIdentifiedCallStack('withEnv', List.class)
+        registerIdentifiedCallStack('dir', String.class)
+    }
+
+    void runCall(String name, Integer idx = 0, Boolean doClear = true) {
+        CallStack callStack = callStacks[name]
+        if(callStack== null) {
+            throw new RuntimeException("No such call stack ${name}")
+        }
+        callStack.run(idx, doClear)
+    }
+
+    public <T> void runIdentified(String name, T value, Boolean doClear = true) {
+        IdentifiedCallStack<Object> callStack = identifiedCallStacks[name]
+        if(callStack== null) {'pipeline'
+            throw new RuntimeException("No such call stack ${name}")
+        }
+        callStack.run(value, doClear)
+    }
+
+    Object runIdentifiedByIndex(String name, Integer idx = 0, Boolean doClear = true) {
+        IdentifiedCallStack<Object> callStack = identifiedCallStacks[name]
+        if(callStack== null) {
+            throw new RuntimeException("No such call stack ${name}")
+        }
+        return callStack.run(idx, doClear)
+    }
+
+    void registerCallStack(String name) {
+        callStacks.put(name, new CallStack(name, base.helper))
+    }
+
+    public <T> void registerIdentifiedCallStack(String name, Class<T> clazz) {
+        identifiedCallStacks.put(name, new IdentifiedCallStack<T>(name, base.helper, clazz))
+    }
+
+    void clearAllCalls() {
+        base.helper.clearCallStack()
+        callStacks.values().forEach { it.clearCalls() }
+        identifiedCallStacks.values().forEach { it.clearCalls() }
+    }
+
+
+    class CallStack {
+        String name
+        PipelineTestHelper helper
+
+        List<Closure> calls = []
+
+        CallStack(name, helper) {
+            this.name = name
+            this.helper = helper
+
+            this.helper.registerAllowedMethod(name, [Closure.class], { calls.add(it) })
```

Comment:
This is a fundamentally different approach than most of this framework, because the closures in the jenkinsfile are not implicitly executed but must be described by  the test itself. (which imo makes it a more focused test)